lib/TPP/ToBlockLayoutAndBack.cpp

```diff
@@ -776,9 +776,169 @@ struct PropagateThroughElementWiseOp
   }
 };
 
+//===----------------------------------------------------------------------===//
+// Passes
+//===----------------------------------------------------------------------===//
+
+// Pack MatmulOp.
+struct DoItOnMatmul : public OpRewritePattern<linalg::MatmulOp> {
+  DoItOnMatmul(MLIRContext *context, ArrayRef<int64_t> blockingFactors,
+               PatternBenefit benefit = 1)
+      : OpRewritePattern<linalg::MatmulOp>(context, benefit),
+        blockingFactors(blockingFactors) {}
+
+  LogicalResult matchAndRewrite(linalg::MatmulOp matmulOp,
+                                PatternRewriter &rewriter) const override {
+    FailureOr<linalg::GenericOp> packedMatmul = mlir::linalgx::packMatmulOp(
+        rewriter, matmulOp,
+        getAsOpFoldResult(rewriter.getI64ArrayAttr(blockingFactors)));
+    if (failed(packedMatmul))
+      return failure();
+    return success();
+  }
+
+private:
+  ArrayRef<int64_t> blockingFactors;
+};
+
+// From linalg.generic to linalg.matmul.
+struct DeGeneralizeMatmul : public OpRewritePattern<linalg::GenericOp> {
+  using OpRewritePattern<linalg::GenericOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(linalg::GenericOp linalgOp,
+                                PatternRewriter &rewriter) const override {
+    if (!linalgOp.hasTensorSemantics())
```

Comment:
We need to add an issue to move this pack to tensors, so we can bring the upstream ops.

After the pass work, this is one of the most important things we need to finish.

Reply:
Sorry did not get here. Line 808 is just a de-linearization pattern.